Allow all classes for @apply

garymathews opened the pull request:
Relates to https://github.com/tailwindlabs/tailwindcss/issues/6321

- Basically, _apply_ only works for Tailwind generated classes under JIT; which is default when moving to Tailwind 3.
  - This issue is also present in Tailwind 2 when using JIT.

garymathews commented on src/lib/expandApplyAtRules.js:
```diff
@@ -122,6 +122,27 @@ function processApply(root, context) {
+          applyClassCache.set(name, [
+            [
+              {
+                sort: BigInt(context.layerOrder.user),
+                layer: 'user',
```
I'm not too sure about these

RobinMalfait commented:
Hey! Thank you for your bug report!
Much appreciated! 🙏

If I run your tests against the current `master` branch, then it already passes. I'll look into the actual issue and mark you as a contributor if I have a fix.

garymathews commented:
@RobinMalfait, thanks for fixing this 👍  https://github.com/tailwindlabs/tailwindcss/pull/6589

However, I noticed it's required to reference `@tailwind;` to the particular css file with the `@apply` for this to work (don't even need to define a directive); where before this was not necessary. (I should only need to define directives in my main css file)

```JS
it('should pickup all classes', () => {
  let config = {
    content: [{ raw: html`<div class="foo"></div>` }],
    plugins: [],
  }

  let input = css`
    @tailwind; /* remove me I break */

    .bop {
      color: red;
    }
    .bar {
      background-color: blue;
    }
    .foo {
      @apply absolute bar bop;
    }
  `

  return run(input, config).then((result) => {
    return expect(result.css).toMatchFormattedCss(css`
      @tailwind;
      .bop {
        color: red;
      }
      .bar {
        background-color: blue;
      }
      .foo {
        position: absolute;
        color: red;
        background-color: blue;
      }
    `)
  })
})
```
```
The `bar` class does not exist. If `bar` is a custom class, make sure it is defined within a `@layer` directive.
```

garymathews commented:
I found its due to an optimization where css files without a directive are ignored:
[setupTrackingContext.js#L127](https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/setupTrackingContext.js#L127)
[setupWatchingContext.js#L244](https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/setupWatchingContext.js#L244)

garymathews committed:
Always include css with apply in context

garymathews commented:
@RobinMalfait I've updated my PR with a solution that maintains the optimization for css without @tailwind or @apply

thecrypticace committed:
Merge branch 'master' into allow-all-classes

# Conflicts:
#	src/lib/setupWatchingContext.js

thecrypticace committed:
Use let

We use it more consistently

thecrypticace committed:
Remove early return

To match the style of the surrounding code

thecrypticace committed:
Don't return layer directives

They do not need to be returned here. If it's needed in the future its easy enough to add it back.

thecrypticace committed:
Use let

thecrypticace committed:
Update changelog

RobinMalfait committed:
fix typo

And re-format comments
